src/search/searchBar.component.tsx

```diff
@@ -12,35 +14,101 @@ import {
   selectSearchParams,
 } from '../state/slices/searchSlice';
 
+export type TimeframeDates = {
+  fromDate: Date | null;
+  toDate: Date | null;
+};
+
 const SearchBar = (): React.ReactElement => {
   const dispatch = useAppDispatch();
 
-  const searchParams = useAppSelector(selectSearchParams);
+  const searchParams = useAppSelector(selectSearchParams); // the parameters sent to the search query itself
   const { dateRange, shotnumRange } = searchParams;
 
-  const [fromDate, setFromDate] = React.useState<Date | null>(
-    dateRange.fromDate ? new Date(dateRange.fromDate) : null
-  );
-  const [toDate, setToDate] = React.useState<Date | null>(
-    dateRange.toDate ? new Date(dateRange.toDate) : null
-  );
+  // ########################
+  // DATE-TIME FIELDS
+  // ########################
+  // The FROM and TO dates sent as part of the query (if any)
+  const [searchParameterFromDate, setSearchParameterFromDate] =
+    React.useState<Date | null>(
+      dateRange.fromDate ? new Date(dateRange.fromDate) : null
+    );
+  const [searchParameterToDate, setSearchParameterToDate] =
+    React.useState<Date | null>(
+      dateRange.toDate ? new Date(dateRange.toDate) : null
+    );
 
-  const [shotnumMin, setShotnumMin] = React.useState<number | undefined>(
-    shotnumRange.min ?? undefined
-  );
-  const [shotnumMax, setShotnumMax] = React.useState<number | undefined>(
-    shotnumRange.max ?? undefined
+  // ########################
+  // TIMEFRAME
+  // ########################
+  // Timeframe range is NOT sent as part of query
+  // Instead, we use this to determine the FROM and TO dates
+  // This is set in the timeframe search box by the user
+  // If the user edits any of the date fields otherwise, this is set to null
+  const [timeframeRange, setTimeframeRange] =
+    React.useState<TimeframeRange | null>(null);
+
+  const calculateTimeframeDateRange = (
+    timeframe: TimeframeRange
+  ): { from: Date; to: Date } => {
+    const to = new Date();
+    const from = new Date();
```

Comment:
I think this is the reason perhaps why you're sometimes off by a second in tests - theoretically between these two lines the seconds could tick over. If we change the definition of `from` to be `const from = new Date(to);`, this ensures that a) what we consider "now" is the same in both `to` and `from` and b) `from` is not a reference to `to` and is instead a new date object with the same time info as `to`

Reply:
@louise-davies okay having gone back over things, I now remember the exact problems. The issue stems from the fact that when I calculate the `expectedDate` in tests and when the test clicks the relative timeframe button (and calculates the current date when THAT happens), those two date objects MAY differ by one second due to happening in two different seconds.

I am unsure how to address this in unit tests.

I attempted to address this in e2e tests through the use of `cy.clock()` and `clock.setSystemTime()` but these are only available in Cypress ^v10.7

So I have not resolved anything in this regard

Reply:
In unit tests you could mock the `Date` constructor to always give the same date if you provide it with zero arguments I suppose

`cy.clock` can be used without `setSystemTime` to mock `Date` functions - see https://github.com/ral-facilities/datagateway/blob/e893e0e49ed5b5b32442b95e29b273e16b3ad764/packages/datagateway-download/cypress/integration/downloadStatus.spec.ts#L39

Reply:
I will give this a go

Reply:
I have managed to get this working with unit tests but mocking the time across e2e tests I haven't. I believe it's because you can mock what Cypress believes the time to be but you still have the problem with the built-and-running application itself still using the current system time